boot/modeenv.go

```diff
@@ -581,3 +581,30 @@ func (s *bootCommandLines) UnmarshalJSON(data []byte) error {
 	*s = bootCommandLines(asList)
 	return nil
 }
+
+// MaybeReadModeenv uses ReadModeenv() with the default root directory, but
+// ignores ENOENT errors and returns a nil, but no error.
+func MaybeReadModeenv() (*Modeenv, error) {
+	modeenv, err := ReadModeenv("")
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("cannot read modeenv: %v", err)
+	}
+	return modeenv, nil
+}
+
+// SystemMode returns the current mode of the system and a flag indicating
+// whether the mode is explicitly set through modeenv.
+func SystemMode() (mode string, explicit bool, err error) {
+	modeenv, err := MaybeReadModeenv()
+	if err != nil {
+		return "", false, err
+	}
+
+	if modeenv != nil {
+		// we have the modeenv, making the system mode explicit
+		return modeenv.Mode, true, nil
+	}
+
+	// lacking the modeenv, system is implicitly in "run" mode
```

Comment:
Is that some legacy? I wonder if we should log this case.

Reply:
Ah or do we have no modeenv on first boot?

Reply:
I thought the default was install? is that for snap-boostrap only? so having a default here is a bit strange, it should be probably passed in instead

Reply:
Install is the default only when we figure out the mode from kernel command line (and `snapd_recovery_mode` is set).

Passed in, as a parameter by the caller?

Reply:
Snap-bootstrap generates an ephemeral modeenv for install, recover and factory-reset. snap-bootstrap decides the mode based on the kernel command line.

Reply:
@bboozzoo 

> Passed in, as a parameter by the caller?

yes, I meant that, I think it's bit clearer if the default is chosen in devicestate

Reply:
If the kernel command line is not set, then snap-bootstrap will do install.

I suppose modeenv is missing only if snap-boostrap is not used.

Reply:
pushed a change for this